[SPARK-5207] [MLLIB] StandardScalerModel mean and variance re-use

ogeagla opened the pull request:
This seems complete, the duplication of tests for provided means/variances might be overkill, would appreciate some feedback.

ogeagla committed:
[SPARK-5207] [MLLIB] [WIP] change StandardScalerModel contructor to not be private to mllib, added tests for newly-exposed functionality

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
ok to test

mengxr commented:
cc @dbtsai

SparkQA commented:
[Test build #25895 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25895/consoleFull) for   PR 4140 at commit [`64408a4`](https://github.com/apache/spark/commit/64408a44f1c67399a26580462a84ea206828e341).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25895 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25895/consoleFull) for   PR 4140 at commit [`64408a4`](https://github.com/apache/spark/commit/64408a44f1c67399a26580462a84ea206828e341).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25895/
Test PASSed.

dbtsai commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -61,20 +61,30 @@ class StandardScaler(withMean: Boolean, withStd: Boolean) extends Logging {
+    val variance: Vector,
+    private var withMean: Boolean = false,
+    private var withStd: Boolean = true) extends VectorTransformer {
 
```
The default argument is not friendly for Java though; why don't we add another constructor which takes only mean and variance?

dbtsai replied:
Also, users will want to know if `withMean` or `withStd` is used, do we really need to have them as private variables?

dbtsai commented:
For the unit-test part, is it possible not to change too much? Also, it will be easier to debug if the assertion is in the test instead of abstract out. For example, having `validateConstant` function is not necessary, probably more easy to read to have all the assert code in the test. 

Having the data as global variables is okay for me. 

Thanks.

ogeagla committed:
[SPARK-5207] [MLLIB] [WIP] make withMean and withStd public, add constructor which uses defaults, un-refactor test class

SparkQA commented:
[Test build #26038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26038/consoleFull) for   PR 4140 at commit [`997d2e0`](https://github.com/apache/spark/commit/997d2e0a3bbfd1be6c0a556393bbcfbd18404f77).
 * This patch merges cleanly.

ogeagla commented:
@dbtsai that makes sense.  I've changed this back in latest commit.

SparkQA commented:
[Test build #26038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26038/consoleFull) for   PR 4140 at commit [`997d2e0`](https://github.com/apache/spark/commit/997d2e0a3bbfd1be6c0a556393bbcfbd18404f77).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StandardScalerModel (`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26038/
Test PASSed.

dbtsai commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -61,20 +61,34 @@ class StandardScaler(withMean: Boolean, withStd: Boolean) extends Logging {
+    var withStd: Boolean) extends VectorTransformer {
 
   require(mean.size == variance.size)
 
```
Moving `require` down to `this` constructor.

dbtsai commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -61,20 +61,34 @@ class StandardScaler(withMean: Boolean, withStd: Boolean) extends Logging {
+    val variance: Vector,
+    var withMean: Boolean,
+    var withStd: Boolean) extends VectorTransformer {
 
```
Moving `require` down to `this` constructor.

dbtsai commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -61,20 +61,34 @@ class StandardScaler(withMean: Boolean, withStd: Boolean) extends Logging {
+  def this(mean: Vector, variance: Vector) {
+    this(mean, variance, false, true)
+  }
+
```
Moving `require` to the bottom of `this` constructor. Also add `@DeveloperApi` annotation to both `setWithMean` and `setWithStd` APIs.

mengxr replied:
I have a question about this API. If the default `withMean` is `false`, why do we require `mean` in the constructor? If the feature dimension is really large, this puts some extra cost that cannot be ignored. Similarly, should we take std directly instead of variance in the constructor? My proposal is the following:

* `StandoardScalerModel(std: Vector, mean: Vector, withStd: Boolean, withMean: Boolean)`.  I put `variance` in front of `mean` because scaling is used more frequently than shifting.
* `this(std: Vector, mean: Vector)`: enable `withMean` and `withStd` based on whether the input arguments are null or not. Throw exception is both are null.
* `this(std: Vector) = this(std, null)`.
* `setWithMean` and `setWithStd` check whether the corresponding `mean`/`variance` is null or not and throw exceptions if a user want to set it to true while the value is null.

dbtsai replied:
Sounds reasonable for me. Although the changes will be larger, this will be more handy and save extra space if withMean is not used.

ogeagla replied:
@mengxr Just to make sure I'm clear, are you suggesting changing the `StandardScalerModel` to take the standard deviation vector (instead of variance)?  Or are you just calling it 'std' for short?

dbtsai replied:
In my opinion, taking `variance` will be ideal since it's the output of `MultivariateOnlineSummarizer`.

dbtsai commented:
LGTM except those two minor details. Thanks.

ogeagla committed:
[SPARK-5207] [MLLIB] [WIP] Incorporate code review feedback: change arg ordering, add dev api annotations, do better null checking, add another test and some doc for this.

SparkQA commented:
[Test build #26224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26224/consoleFull) for   PR 4140 at commit [`9078fe0`](https://github.com/apache/spark/commit/9078fe049b40c67694c5cef63ce943e158079d99).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26224/consoleFull) for   PR 4140 at commit [`9078fe0`](https://github.com/apache/spark/commit/9078fe049b40c67694c5cef63ce943e158079d99).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StandardScalerModel (`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26224/
Test FAILed.

ogeagla commented:
@mengxr I've incorporated your comments.

@dbtsai I ran lint-scala and it fails due to files being too big, and that's what's failing the build.  Shall I separate the `StandardScaler` and `StandardScalerModel` impl and tests into separate files?

Also, I added a tiny bit to the docs.  I would be happy to add more if you think it's appropriate.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.rdd.RDD
 /**
  * :: Experimental ::
- * Standardizes features by removing the mean and scaling to unit variance using column summary
+ * Standardizes features by removing the mean and scaling to unit standard deviation using column summary
```
line too wide. the limit is 100 chars.

mengxr commented:
@ogeagla You can see the error messages from the Jenkins build results (https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26224/), where it shows

~~~
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:29: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:78: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:79: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:93: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:117: File line length exceeds 100 characters
[error] (mllib/compile:scalastyle) errors exist
[error] Total time: 5 s, completed Jan 28, 2015 4:53:33 AM
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:29: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:78: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:79: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:93: File line length exceeds 100 characters
[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:117: File line length exceeds 100 characters
[error] (mllib/compile:scalastyle) errors exist
~~~

So it is the line length but not the file length.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/feature/StandardScaler.scala:
```diff
@@ -61,19 +61,39 @@ class StandardScaler(withMean: Boolean, withStd: Boolean) extends Logging {
+    this
+  }
 
   private lazy val factor: Array[Double] = {
```
Back to the discussion on using `std` instead of `variance`, Essentially we are computing the 1/`std` here. If we want to save the model, we certainly prefer saving `factor` to saving `variance`, so we don't need to recompute this while loading it back. It also saves storage. If we have `std`, we don't need to create `factor` and the transformation can be done on the fly with `std`. The overhead is just a `if` check.

ogeagla replied:
I agree with that, I can make those changes.  An additional one-time overhead is the computing of sqrt of the `variance` from the `summary` in `StandardScaler.fit` to provide to the `StandardScalerModel` constructor.

ogeagla committed:
[SPARK-5207] [MLLIB] [WIP] change StandardScalerModel to take stddev instead of variance

SparkQA commented:
[Test build #26472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26472/consoleFull) for   PR 4140 at commit [`fa64dfa`](https://github.com/apache/spark/commit/fa64dfaa9f27dc6291d76e8e747ccd87bf3c8d2e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26472/consoleFull) for   PR 4140 at commit [`fa64dfa`](https://github.com/apache/spark/commit/fa64dfaa9f27dc6291d76e8e747ccd87bf3c8d2e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StandardScalerModel (`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26472/
Test PASSed.

mengxr commented:
LGTM. Merged into master. Thanks!
